### PR TITLE
Added metadata parser

### PR DESCRIPTION
**Pull request by aasim**

### CODE REVIEW

File: src/Microsoft.Sbom.Extensions/ISbomParser.cs [hunk 1]
```diff
@@ -58,7 +58,7 @@ public interface ISbomParser
     /// </summary>
     /// <param name="stream"></param>
     /// <returns></returns>
-    SBOMMetadata GetMetadata();
+    Spdx22Metadata GetMetadata();
```
[the remark this answers is not in the document]
Comment (ryanbrandenburg) [reply src/Microsoft.Sbom.Extensions/ISbomParser.cs:61]: Yeah, maybe Spdx22Metadata should extend from an `ISBOMMetadata` or something?

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/SPDXParser.cs [hunk 2]
```diff
@@ -159,8 +168,68 @@ ParserState MoveToNextState()
...
+                    metadata.SpdxVersion = nextTokenString;
+                    break;
+                case Constants.DataLicenseHeaderName:
+                    metadata.DataLicense = nextTokenString;
+                    break;
+                case Constants.DocumentNameHeaderName:
+                    metadata.Name = nextTokenString;
+                    break;
+                case Constants.DocumentNamespaceHeaderName:
+                    metadata.DocumentNamespace = new Uri(nextTokenString);
```
Comment (ryanbrandenburg) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/SPDXParser.cs:198]: GitHub thinks this might NullRef.

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/SPDXParser.cs [hunk 3]
```diff
@@ -159,8 +168,68 @@ ParserState MoveToNextState()
...
+        return internalParserState;
+
+        ParserState ProcessPropertyInternal()
+        {
+            var reader = new Utf8JsonReader(buffer, isFinalBlock: isFinalBlock, readerState);
+            
+            switch (currentRootPropertyName)
+            {
+                case Constants.SPDXVersionHeaderName:
+                    metadata.SpdxVersion = nextTokenString;
```
Comment (ryanbrandenburg) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/SPDXParser.cs:189]: Why do this instead of `ParserUtils.GetString()` or whatever? Whatever the reason is we should codify it in a comment for future generations.

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/SPDXParser.cs [hunk 4]
```diff
@@ -287,12 +361,12 @@ public IEnumerable<SBOMRelationship> GetRelationships()
             throw new ParserException($"The parser is not currently enumerating relationships. Current state: {CurrentState}");
         }
 
-        while (GetPackages(stream, out SPDXRelationship sbomRelationship) != 0)
+        while (GetRelationships(stream, out SPDXRelationship sbomRelationship) != 0)
```
Comment (ryanbrandenburg) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/SPDXParser.cs:364]: lol thanks, I noticed this and got confused while reading the class.

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Utils/ParserStateResult.cs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+﻿using Microsoft.Sbom.Contracts.Enums;
+
+namespace Microsoft.Sbom.Utils;
+
+#pragma warning disable SA1313 // Parameter names should begin with lower-case letter
```
Comment (ryanbrandenburg) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Utils/ParserStateResult.cs:5]: I think I'd rather change the EditorConfig preferences, but we can do that in a separate PR, no need to block on this.

File: src/Microsoft.Sbom.Extensions/ISbomParser.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (aasim) [reply src/Microsoft.Sbom.Extensions/ISbomParser.cs:61]: Yeah that was my bigges pet peeve about this. This should technically be agnostic to the format... however, the metadata included is very much tied to spdx22, so couldn't come up with a better name.

In the future, if spdx3 comes around, we should create a base class with the common properties, and spdx22metadata and spdx30metadata classes with additional properties. The API then should return the base class. 

Its not SBOMMetadata, since that has a different meaning in our world, its used for generation.

[the remark this answers is not in the document]
Comment (ryanbrandenburg) [reply src/Microsoft.Sbom.Extensions/ISbomParser.cs:61]: Hmm, maybe we could call this `GetSpdx22Metadata` then to future-proof ourselves?

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/SPDXParser.cs [hunk 6]
```diff
@@ -1,6 +1,8 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (ryanbrandenburg) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/SPDXParser.cs:4]: The copyright got doubled somehow.

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/SPDXParser.cs [hunk 7]
```diff
@@ -159,8 +172,77 @@ ParserState MoveToNextState()
...
+            readerState = reader.CurrentState;
+            currentRootPropertyName = result.PropertyName;
+            nextTokenString = result.NextToken;
+            return result.State;
+        }
+    }
+
     private void ValidateParsingComplete()
     {
+        if (!metadataStateProcessed)
```
Comment (ryanbrandenburg) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/SPDXParser.cs:241]: NitPick: I don't think this actually validates anything important to the parser. As far as I can tell it's just mandating that  the user calls `GetMetadata`. That makes sense for `GetReferences`, `GetPackages`, etc because they can control the state of the stream/parser, but the metadata is already parsed regardless, it's just a matter of whether the user views the result or not.

**Review by ryanbrandenburg — APPROVED** [review]

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/SPDXParser.cs (hunk 7 above)
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/SPDXParser.cs:241]: The way the state machine is designed, the only way to get to the finished state without a failure event is to get the metadata. Since we want that to be called last. So this becomes necessary.

File: src/Microsoft.Sbom.Extensions/ISbomParser.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (edgarrs) [reply src/Microsoft.Sbom.Extensions/ISbomParser.cs:61]: I would have made this method generic to return T. The entire interface would be ISbomParser<T> where T will be Spdx22Metadata  for this specific implementation. We will need to add ISbomMetadata interface or something to restrict T.
